receiver/prometheusreceiver/internal/metricsbuilder.go

```diff
@@ -0,0 +1,549 @@
+		if b.currentMetric != nil {
+			if err := b.completeCurrentMetric(); err != nil {
+				return err
+			}
+		}
+
+		// setup new metric group
+		metadata, ok := b.mc.Metadata(metricFamily)
+
+		// perform a 2nd lookup with the original metric name
+		// this can only happen if there's a metric is not histogram or summary and ends with one of the suffixes
+		// not a good name, but you can still do it that way. hence perform a 2nd lookup with original name
+		if !ok && metricName != metricFamily {
+			metricFamily = metricName
+			metadata, ok = b.mc.Metadata(metricFamily)
+			// still not found, this can happen for metrics has no TYPE HINTS
+			if !ok {
+				metadata.Metric = metricFamily
+				metadata.Type = textparse.MetricTypeUnknown
+			}
+		}
+		b.currentMetricFamily = metricFamily
+		b.currentMetadata = &metadata
+		b.currentDpGroups = make(map[string][]*dataPoint)
+		b.currentDpGOrder = make(map[string]int)
+		b.currentMetadataGroups = make(map[string]*dataPointGroupMetadata)
+		b.currentMetric = &metricspb.Metric{
+			MetricDescriptor: &metricspb.MetricDescriptor{
+				Name:        metricFamily,
+				Description: metadata.Help,
+				Unit:        heuristicalMetricAndKnownUnits(metricFamily, metadata.Unit),
+				Type:        convToOCAMetricType(metadata.Type),
+				// Due to the fact that scrapeLoop strips any tags with emtpy value, we cannot get the
+				// full set of labels by just looking at the first metrics, thus setting the LabelKeys of
+				// this proto need to be done after finished the whole group
+			},
+			Timeseries: make([]*metricspb.TimeSeries, 0),
+		}
+		b.currentMetricLabelKeys = make(map[string]int)
+	}
+	return nil
+}
+
+func (b *metricBuilder) updateLabelKeys(ls labels.Labels) {
+	for _, l := range ls {
+		if isUsefulLabel(b.currentMetric.MetricDescriptor.Type, l.Name) {
+			if _, ok := b.currentMetricLabelKeys[l.Name]; !ok {
+				b.currentMetricLabelKeys[l.Name] = len(b.currentMetricLabelKeys)
+			}
+		}
+	}
+}
+
+func (b *metricBuilder) getLabelKeys() []*metricspb.LabelKey {
+	lks := make([]*metricspb.LabelKey, len(b.currentMetricLabelKeys))
+
+	for k, v := range b.currentMetricLabelKeys {
+		lks[v] = &metricspb.LabelKey{Key: k}
+	}
+
+	return lks
+}
+
+func (b *metricBuilder) getLabelValues(ls *labels.Labels) []*metricspb.LabelValue {
+	lvs := make([]*metricspb.LabelValue, len(b.currentMetricLabelKeys))
+	for k, v := range b.currentMetricLabelKeys {
+		value := ls.Get(k)
+		lvs[v] = &metricspb.LabelValue{Value: value, HasValue: value != ""}
+	}
+
+	return lvs
+}
+
+func (b *metricBuilder) completeCurrentMetric() error {
+	switch mtype := b.currentMetric.MetricDescriptor.Type; mtype {
+	case metricspb.MetricDescriptor_GAUGE_DISTRIBUTION, metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		groups := b.currentDpGroupOrdered()
+		if len(groups) == 0 {
+			// this can happen if only sum or count is added, but not following data points
+			return errors.New("no data point added to summary")
```

Comment:
For Python, the Prometheus client code is incomplete for summaries and only exports sum and count - consider changing this to a log message and returning nil otherwise the entire scrape will be lost for any Python application with a Summary metric.

Reply:
Good catch, I have observed similar behavior very recently from a MicroMeter' prometheus endpoint that a summary only contains sum and count. 
Returning nil, as suggested, is one way to solve the issue, however, the down side is that we might loss some important metrics which customers are interested in.  Or we can convert these kind of metrics differently into something like  two counters, or dose OpenCensus format allow to have Summary Snapshots with no Percentile values?

Reply:
Additionally, the Prometheus Go client now exports empty summaries by default.

I believe that exporting summary snapshot with empty percentile values makes the most sense.

@songy23 @rghetia - wdyt?

Reply:
I agree with empty percentile values.

Reply:
will do.   I am also going to drop gaugehistogram support from this pr until an official spec of this data type is provided.  as hint by @dinooliva , I have checked some code from the python prometheus client, and found it had some code related to this `gaugehistogram` type from this PR https://github.com/prometheus/client_python/pull/306 , and it seems like its a bit different from regular histograms, instead of using _count/_sum it uses  _gcount/_gsum as postfixes. the PR has also attached a draft spec link to openmetrics, the next of prometheus, which can be found from https://docs.google.com/document/d/1KwV0mAXwwbvvifBvDKH_LU1YjyXE_wxCkHNoCGq1GX0/edit#heading=h.1cvzqd4ksd23 

it also has some interesting new features like _created to indicate the start time of metrics, however, this spec is still in draft stage, and I am not sure when it will be published.